packages/devtools_shared/test/utils/file_utils_test.dart

```diff
@@ -2,81 +2,183 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:io';
+
 import 'package:devtools_shared/devtools_shared.dart';
+import 'package:dtd/dtd.dart';
+import 'package:path/path.dart' as p;
 import 'package:test/test.dart';
 
+import '../helpers.dart';
+
 void main() {
   group('file uri helpers', () {
-    test('rootFromFileUriString', () {
-      // Dart file under 'lib'
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/lib/main.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/lib/sub/main.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
+    TestDtdConnectionInfo? dtd;
+    DartToolingDaemon? testDtdConnection;
 
-      // Dart file under 'bin'
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/bin/script.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/bin/sub/script.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
+    const projectRootParts = ['absolute_path_to', 'my_app_root'];
+    late String projectRoot;
 
-      // Dart file under 'test'
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/test/some_test.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/test/sub/some_test.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
+    late File libFile;
+    late File libSubFile;
+    late File binFile;
+    late File binSubFile;
+    late File testFile;
+    late File testSubFile;
+    late File integrationTestFile;
+    late File integrationTestSubFile;
+    late File benchmarkFile;
+    late File benchmarkSubFile;
+    late File exampleFile;
+    late File exampleSubFile;
+    late File anyFile;
+    late File anySubFile;
 
-      // Dart file under 'integration_test'
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/integration_test/some_test.dart',
-        ),
-        equals('file:///Users/me/foo/my_app_root'),
-      );
-      expect(
-        packageRootFromFileUriString(
-          'file:///Users/me/foo/my_app_root/integration_test/sub/some_test.dart',
+    setUp(() async {
+      dtd = await startDtd();
+      expect(dtd!.uri, isNotNull, reason: 'Error starting DTD for test');
+      testDtdConnection = await DartToolingDaemon.connect(Uri.parse(dtd!.uri!));
+
+      // Set up the project root.
+      final tmpDirectory = Directory.systemTemp.createTempSync();
```

Comment:
Consider adding a comment describing the directory structure. It's hard to visualize just based on the code.